canonical/main.go

```diff
@@ -0,0 +1,377 @@
+		}
+
+		// Don't need to do anything with directory headers.
+		if hdr.FileInfo().IsDir() {
+			tw.WriteHeader(hdr)
+			continue
+		}
+
+		// Only process .html files.
+		if !strings.HasSuffix(hdr.Name, ".html") {
+			tw.WriteHeader(hdr)
+			if l, err := io.Copy(tw, tr); err != nil || l != hdr.Size {
+				return fmt.Errorf("error writing %q %q: wrote %v bytes: %v", name, hdr.Name, l, err)
+			}
+			continue
+		}
+
+		// Scanner to read every line.
+		scanner := bufio.NewScanner(tr)
+		// Write to a buffer so we can get the new number of bytes for the record.
+		out := &bytes.Buffer{}
+
+		// foundCanonical and inHead keep track of where we are, naively,
+		// in the HTML file.
+		foundCanonical := false
+		inHead := false
+
+		filename := hdr.Name[len("./"):]
+
+		// Note: this does not preserve line endings.
+		// You can `diff --ignore-space-change` to verify there
+		// are no changes to content.
+		for scanner.Scan() {
+			// Read as bytes to avoid an unnecessary string allocation.
+			text := scanner.Bytes()
+
+			// Trim space rather than use regex.
+			trimmed := bytes.TrimSpace(text)
+			if bytes.Equal(trimmed, startHead) {
+				inHead = true
+			}
+
+			if inHead && bytes.Contains(text, canonical) {
+				foundCanonical = true
+			}
+
+			if inHead && bytes.Equal(trimmed, endHead) {
+				if !foundCanonical {
+					link, err := canonicalLink(ctx, pkg, filename)
+					if err != nil {
+						return fmt.Errorf("canonicalLink: %v", err)
+					}
+					fmt.Fprintf(out, "    <link rel=\"canonical\" href=%q/>\n", link)
+					foundCanonical = true
+				}
+				inHead = false
+			}
+
+			out.Write(text)
+			fmt.Fprintln(out) // Text does not have a newline. Add one.
+		}
+		if err := scanner.Err(); err != nil {
+			return fmt.Errorf("error scanning %q %q: %v", name, hdr.Name, err)
+		}
+
+		// Update the header to have the right size. Otherwise, leave hdr unchanged.
+		hdr.Size = int64(out.Len())
+		tw.WriteHeader(hdr)
+		if l, err := tw.Write(out.Bytes()); err != nil || l != int(hdr.Size) {
+			return fmt.Errorf("error writing %q %q: wrote %v bytes: %v", name, hdr.Name, l, err)
+		}
+
+		atomic.AddInt64(&numFiles, 1)
+		bar.Set("prefix", fmt.Sprintf("Canonicalizing... %d files | ", numFiles))
+	}
+
+	if err := tw.Close(); err != nil {
+		return fmt.Errorf("error closing tar %v: %v", name, err)
+	}
+	if err := gw.Close(); err != nil {
+		return fmt.Errorf("error closing gzip %v: %v", name, err)
+	}
+	if err := ow.Close(); err != nil {
+		return fmt.Errorf("error closing obj %q writer: %v", name, err)
+	}
+
+	return nil
+}
+
+// parsePkg gets the package name from the tarball name.
+func parsePkg(tarballName string) string {
+	name := tarballName[len(dotnetPrefix):] // Trim docfx- prefix.
+	name = name[:strings.Index(name, "-")]  // Discard after first remaining -.
+	return name
+}
+
+// canonicalLink returns the canonical link for the given package and page.
+func canonicalLink(ctx context.Context, pkg, page string) (string, error) {
```

Comment:
Just out of curiosity, why do we need a whole separate service to provide these links?

Reply:
It's relatively complex logic, which we need in the .NET side of things as well (for generating canonical links in googleapis.dev in new packages, and for setting redirects in github.com). A Cloud Function is a really simple way of sharing this logic.